### PR TITLE
training argument for tf.nn.rnn_cell.DropoutWrapper

### PR DESCRIPTION
**Pull request by resec**
This PR resolves issue #9775 

Added a `training` argument to  `tf.nn.rnn_cell.DropoutWrapper`, identical to `tf.layers.dropout`.

Also added to testcase to `tf.contrib.rnn.python.kernel_tests/core_rnn_cell_test.py` for this new argument.

### CODE REVIEW

**Comment by ebrevdo** [comment]
We're in the process of fixing a bug in dropout which makes some changes to these lines of code.  Please bear with us a day or two and you may need to resync once that bug is fixed.

**Comment by resec** [comment]
@ebrevdo I merge your bugfix code for state dropout, please let me know if I break anything. My main idea is if `training` is marked `True`, then all dropout intention should be suppressed, so I just move those bugfix lines into the state dropout sub function. Thanks.

**Comment by drpngx** [comment]
@ebrevdo what are you thoughts on this?

**Comment by sb2nov** [comment]
@ebrevdo ping

File: tensorflow/python/ops/rnn_cell_impl.py [hunk 1]
```diff
@@ -649,7 +650,7 @@ class DropoutWrapper(RNNCell):
 
   def __init__(self, cell, input_keep_prob=1.0, output_keep_prob=1.0,
```
Comment (ebrevdo) [inline tensorflow/python/ops/rnn_cell_impl.py]: the `training` argument should be passed to `call`, not `__init__`.  See, e.g., the `Dropout` [layer impl](https://github.com/tensorflow/tensorflow/blob/r1.3/tensorflow/python/layers/core.py#L257).

**Review by ebrevdo — CHANGES_REQUESTED** [review]

File: tensorflow/python/ops/rnn_cell_impl.py (hunk 1 above)
Comment (resec) [reply tensorflow/python/ops/rnn_cell_impl.py]: 👌  let me update it. thanks.

Comment (resec) [reply tensorflow/python/ops/rnn_cell_impl.py]: sorry, second thought, is it not easier for user define or control the cell action via a `training` in `init`?

**Comment by martinwicke** [comment]
@resec any update?

**Comment by resec** [comment]
@martinwicke sorry, I did not get what @ebrevdo means, I will check more on the implementation but just too busy for the next few days. Will update soon. Thanks.

**Comment by resec** [comment]
@ebrevdo do you mean that I should create a wrapping function which returns the output tensor, to class `tf.nn.rnn_cell.DropoutWrapper`, just like function `tf.layers.dropout` to class `tf.layers.Dropout`?

But normally we pass the created RNN Cell object to other wrapper methods like `tf.nn.dynamic_rnn`. Maybe we should define the training in Cell construction and adjust its actions when is applied?

Please correct me if I am misunderstanding anything.

**Comment by drpngx** [comment]
@resec could you pull rebase and push again?

@ebrevdo any more comment?

**Comment by tensorflowbutler** [comment]
Nagging Assignee: It has been 14 days with no activity and this issue has an assignee. Please update the label and/or status accordingly.

**Comment by tensorflowbutler** [comment]
Nagging Assignee: It has been 14 days with no activity and this issue has an assignee. Please update the label and/or status accordingly.

**Comment by ebrevdo** [comment]
I believe you're right.

File: tensorflow/python/ops/rnn_cell_impl.py [hunk 2]
```diff
@@ -1036,23 +1042,35 @@ def _should_dropout(p):
...
+        # which ones to keep.
+        shallow_filtered_substructure = nest.get_traverse_shallow_structure(
+            self._dropout_state_filter, new_state)
+        return self._dropout(new_state, "state",
+                             self._recurrent_state_noise,
+                             self._state_keep_prob,
+                             shallow_filtered_substructure)
+      new_state = utils.smart_cond(self._training,
+                                   dropped_new_state,
+                                   lambda: array_ops.identity(new_state))
```
Comment (ebrevdo) [inline tensorflow/python/ops/rnn_cell_impl.py]: do you need the `array_ops.identity` ?

File: tensorflow/python/ops/rnn_cell_impl.py [hunk 3]
```diff
@@ -1036,23 +1042,35 @@ def _should_dropout(p):
...
     if _should_dropout(self._output_keep_prob):
-      output = self._dropout(output, "output",
+      def dropped_output():
+        return self._dropout(output, "output",
                              self._recurrent_output_noise,
                              self._output_keep_prob)
+      output = utils.smart_cond(self._training,
+                                dropped_output,
+                                lambda: array_ops.identity(output))
```
Comment (ebrevdo) [inline tensorflow/python/ops/rnn_cell_impl.py]: ditto

**Review by ebrevdo — CHANGES_REQUESTED** [review]

**Comment by protoget** [comment]
@resec, will you try addressing the comments?

File: tensorflow/python/ops/rnn_cell_impl.py (hunk 2 above)
Comment (resec) [reply tensorflow/python/ops/rnn_cell_impl.py]: In `tf.layers.dropout`, it applies `tf.identity` to `inputs` before returning, but I haven't able to fully understand why it has to be this way. I think (guess) it is to make sure the `input` tensor to be in the same device as the `dropped input (if applied)`.

In here, If I am guessing above correctly, `tf.identity` can be skipped as the `scope` or `device` etc. is decided only when the Cell object is applied?

I think `tf.identity` is a little bit difficult to understand, please correct me if I am wrong, thanks.

Comment (resec) [reply tensorflow/python/ops/rnn_cell_impl.py]: @ebrevdo any comment is very appreciated.

**Comment by tensorflowbutler** [comment]
Nagging Assignee @ebrevdo: It has been 14 days with no activity and this issue has an assignee. Please update the label and/or status accordingly.

**Comment by tensorflowbutler** [comment]
It has been 14 days with no activity and the `awaiting response` label was assigned. Is this PR still valid? Assigning the `stalled` label. Please comment to reassure me that this is still being worked on.

**Comment by tensorflowbutler** [comment]
It has been 43 days that this pull-request has stalled. Please create a new pull-request with the requested changes.
